src/rosdiscover/recover/tool.py

```diff
@@ -1,20 +1,331 @@
+        if files.isdir(build_dir):
+            return build_dir
+
+        build_dir = os.path.join(workspace, "build_isolated")
+        if files.isdir(build_dir):
+            return build_dir
+
+        raise ValueError(f"unable to find build directory in workspace: {workspace}")
+
+    def _detect_build_tool(
+        self,
+        workspace: str,
+        build_directory: t.Optional[str] = None,
+    ) -> RosBuildTool:
+        """Detects the build tool that was used to construct a given workspace.
+
+        Parameters
+        ----------
+        workspace: str
+            The absolute path to the workspace.
+        build_directory: str, optional
+            The absolute path to the build directory within the given workspace.
+
+        Raises
+        ------
+        ValueError
+            if the build directory could not be found inside the workspace
+        ValueError
+            if the build tool used to construct the workspace was unrecognized
+        ValueError
+            if the workspace does not provide a .built_by file inside its build directory
+        """
+        assert self._app_instance
+        files = self._app_instance.files
+
+        if not build_directory:
+            build_directory = self._find_build_directory(workspace)
+
+        built_by_path = os.path.join(build_directory, ".built_by")
+
+        try:
+            build_tool_name = files.read(built_by_path, binary=False)
+        except FileNotFoundError:
+            raise ValueError(f"unable to find expected .built_by file: {built_by_path}")
+
+        return RosBuildTool.from_built_by(build_tool_name)
+
+    def _find_compile_commands_file(self, package: roswire.common.Package) -> str:
+        """Locates the compile_commands.json for a given package.
+
+        Raises
+        ------
+        FileNotFoundError
+            if no compile_commands.json was found for the given package
+        ValueError
+            if the build directory could not be found inside the workspace
+        ValueError
+            if the build tool used to construct the workspace was unrecognized
+        ValueError
+            if the workspace does not provide a .built_by file inside its build directory
+
+        Returns
+        -------
+        str
+            the absolute path of the compile_commands.json file
+        """
+        assert self._app_instance
+        files = self._app_instance.files
+        workspace = self._find_package_workspace(package)
+        build_directory = self._find_build_directory(workspace)
+        build_tool = self._detect_build_tool(workspace, build_directory)
+
+        compile_commands_directory: str
+        if build_tool == RosBuildTool.CATKIN_MAKE:
+            compile_commands_directory = build_directory
+        elif build_tool in (RosBuildTool.CATKIN_TOOLS, RosBuildTool.CATKIN_MAKE_ISOLATED):
+            compile_commands_directory = os.path.join(build_directory, package.name)
+        else:
+            assert (
```

Comment:
This assertion doesn't fail or print the message as far as I can see. Do you need `False` as the second argument?

Reply:
I'm mixing up Python and C++ :-) We can either do `assert False, msg` or `raise AssertionError(msg)`

Reply:
I think raise AssertionError, or ValueError.